sdks/java/io/elasticsearch-tests/elasticsearch-tests-common/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTestCommon.java

```diff
@@ -302,4 +303,129 @@ void testWriteWithMaxBatchSizeBytes() throws Exception {
+   * work to the ES server and even passing moderate number of docs can overload the bulk queue and
+   * workers. The post explains more
+   * https://www.elastic.co/blog/why-am-i-seeing-bulk-rejections-in-my-elasticsearch-cluster.
+   * Therefore limit to a small number of docs to test routing behavior only.
+   */
+  void testWriteWithIndexFn() throws Exception {
+    long docsPerScientist = 10; // very conservative
+    long adjustedNumDocs = docsPerScientist * FAMOUS_SCIENTISTS.length;
+
+    List<String> data =
+        ElasticSearchIOTestUtils.createDocuments(
+            adjustedNumDocs, ElasticSearchIOTestUtils.InjectionMode.DO_NOT_INJECT_INVALID_DOCS);
+    pipeline
+            .apply(Create.of(data))
+            .apply(
+                    ElasticsearchIO.write()
+                            .withConnectionConfiguration(connectionConfiguration)
+                            .withIndexFn(new ExtractScientistFn()));
+    pipeline.run();
+
+    // verify counts on each index
+    for (String scientist : FAMOUS_SCIENTISTS) {
+      String index = scientist.toLowerCase();
+
+      // flush writes server side to ensure ready for reading
+      restClient.performRequest("POST", String.format("/%s/_refresh", index));
+
+      String endPoint = String.format("/%s/%s/_search", index, connectionConfiguration.getType());
+      Response response = restClient.performRequest("GET", endPoint);
+      JsonNode searchResult = parseResponse(response);
+      int count = searchResult.path("hits").path("total").asInt();
+      assertEquals(scientist + " index holds incorrect count", docsPerScientist, count);
+    }
+  }
+
+  /**
+   * Returns TYPE_0 or TYPE_1 based on the modulus of the document id.
+   */
+  static class TypeIdFn implements Write.FieldExtractFn {
+    @Override
+    public String apply(JsonNode input) {
+      return "TYPE_" + input.path("id").asInt() % 2;
+    }
+  }
+
+  /**
+   * Tests that documents are dynamically routed to different types. Documents should be routed to
+   * the a type of type_0 or type_1 using modulus approach of the explicit id.
+   */
+  void testWriteWithTypeFn() throws Exception {
+    // defensive coding: this test requires an even number of docs
+    long adjustedNumDocs = (numDocs & 1) == 0 ? numDocs : numDocs + 1;
+
+    List<String> data =
+        ElasticSearchIOTestUtils.createDocuments(
+            adjustedNumDocs, ElasticSearchIOTestUtils.InjectionMode.DO_NOT_INJECT_INVALID_DOCS);
+    pipeline
+        .apply(Create.of(data))
+        .apply(
+            ElasticsearchIO.write()
+                .withConnectionConfiguration(connectionConfiguration)
+                .withTypeFn(new TypeIdFn()));
+    pipeline.run();
+
+    for (int i = 0; i < 2; i++) {
+      // flush writes server side to ensure ready for reading
+      restClient.performRequest(
+          "POST", String.format("/%s/_refresh", connectionConfiguration.getIndex()));
```

Comment:
same as above, use the utility method. And then inline the loop

Reply:
Written this way for the same reason as above - the utility methods use the type and index from the configuration.  Here we are using different URLs for ES.  I could modify the util methods to allow optional overwriting of the index and type if you prefer?